CASSANDRA-18511 Add support for JMX in the in-jvm dtest framework

JeetKunDoug opened the pull request:
Additionally, while testing these changes this commit found and fixed some
JDK17 build/run issues.

JeetKunDoug committed:
CASSANDRA-18511 Add support for JMX in the in-jvm dtest framework

Additionally, while testing these changes this commit found and fixed some
JDK17 build/run issues.

JeetKunDoug committed:
Use SNAPSHOT build of dtest-api and update Circle config

frankgh commented on src/java/org/apache/cassandra/utils/JMXServerUtils.java:
```diff
@@ -362,5 +365,24 @@ public String[] list() throws RemoteException {
+            {
+                UnicastRemoteObject.unexportObject(this, true);
+            }
+            catch (NoSuchObjectException e)
```
maybe rename the variable to indicate it's ignored
```suggestion
            catch (NoSuchObjectException ignored)
```

frankgh commented on src/java/org/apache/cassandra/config/CassandraRelevantProperties.java:
```diff
@@ -219,6 +219,8 @@
 
     /** This property indicates whether disable_mbean_registration is true */
     IS_DISABLED_MBEAN_REGISTRATION("org.apache.cassandra.disable_mbean_registration"),
+    /** This property indicates if the code is running under the in-jvm dtest framework */
```
does the order of the enum matter here? Some enums might be sensitive to the order, so usually new enums are added at the end. It might not be the case here

JeetKunDoug replied:
Thankfully no - this enum is really just a convenience wrapper around system properties, so the ordering doesn't matter (and other folks seem to add things where they make sense vs. at the end as well).

JeetKunDoug replied:
On rebase, it ends up it now needs to be maintained in alphabetical order, so I had to move things around anyway.

frankgh commented on build.xml:
```diff
@@ -263,6 +265,8 @@
         <string>--add-opens java.base/java.lang.reflect=ALL-UNNAMED</string>
         <string>--add-opens java.base/java.net=ALL-UNNAMED</string>
+
+        <string>--add-opens java.rmi/sun.rmi.transport.tcp=ALL-UNNAMED</string>
```
is this `--add-opens` scoped to the in-jvm dtests only?

JeetKunDoug replied:
Neither is scoped specifically to in-jvm dtests, but they aren't in any production startup scripts either... it's only for IDE support and running with `ant test-run` which is documented as `Not for production use!` so I think it's appropriate to have them here.

frankgh commented on build.xml:
```diff
@@ -207,6 +207,7 @@
         <string>--add-exports java.management.rmi/com.sun.jmx.remote.internal.rmi=ALL-UNNAMED</string>
         <string>--add-exports java.rmi/sun.rmi.registry=ALL-UNNAMED</string>
         <string>--add-exports java.rmi/sun.rmi.server=ALL-UNNAMED</string>
+        <string>--add-exports java.rmi/sun.rmi.transport.tcp=ALL-UNNAMED</string>
```
is this `--add-opens` scoped to the in-jvm dtests only?

frankgh commented on src/java/org/apache/cassandra/utils/MBeanWrapper.java:
```diff
@@ -170,6 +186,125 @@ public void unregisterMBean(ObjectName mbeanName, OnException onException)
+            mbs = MBeanServerFactory.createMBeanServer(hostname + "-" + id);
+        }
+
+        public void registerMBean(Object obj, ObjectName mbeanName, OnException onException)
```
can we add the `@Override` annotation to distinguish between internal methods and overridden methods?

JeetKunDoug replied:
Yeah - none of the other existing ones had `@Override` annotations before so I didn't add them here... I can just add them to everything so it's at least consistent.

JeetKunDoug replied:
So added the `@Override` annotations and looking at the result, it just adds a ton of noise to the file (and the diff) - will discuss in person and we can figure out which way we want to go here.

JeetKunDoug replied:
Decided to leave off the `@Override`s for now as it would spam the review with noise. If desired, we could do a follow-up to enforce the code style on files not currently following it.

frankgh commented on src/java/org/apache/cassandra/utils/MBeanWrapper.java:
```diff
@@ -170,6 +186,125 @@ public void unregisterMBean(ObjectName mbeanName, OnException onException)
+                    if (!name.getCanonicalName().contains("MBeanServerDelegate"))
+                        mbs.unregisterMBean(name);
+                } catch (Throwable e) {
+                    logger.debug("Could not unregister mbean {}", name.getCanonicalName());
```
log at warn maybe?
```suggestion
                    logger.warn("Could not unregister mbean {}", name.getCanonicalName(), e);
```

JeetKunDoug replied:
The in-jvm dtest enables debug logging everywhere anyway, and I don't actually think this is a `warn`-level thing as the whole JMX infrastructure gets shut down eventually anyway.

frankgh replied:
should we still log `e` here? or will it create too much output

JeetKunDoug replied:
I don't think it's necessary - it generally happens because it's already been unregistered elsewhere (so it's actually not a huge problem, was more there for, well, debugging), and it would be super-verbose.

frankgh commented on src/java/org/apache/cassandra/utils/MBeanWrapper.java:
```diff
@@ -170,6 +186,125 @@ public void unregisterMBean(ObjectName mbeanName, OnException onException)
+        public MBeanWrapper getDelegate()
+        {
+            return delegate;
+        }
```
new line after brace?
```suggestion
        }

```

JeetKunDoug replied:
Done

frankgh commented on src/java/org/apache/cassandra/utils/RMIClientSocketFactoryImpl.java:
```diff
@@ -0,0 +1,45 @@
+
+/**
+ * This class is used to override the local address the JMX client calculates when trying to connect,
+ * which can otherwise be influences by the system property "java.rmi.server.hostname" in strange and
```
```suggestion
 * which can otherwise be influenced by the system property "java.rmi.server.hostname" in strange and
```

frankgh commented on test/distributed/org/apache/cassandra/distributed/impl/INodeProvisionStrategy.java:
```diff
@@ -56,6 +56,11 @@ public int nativeTransportPort(int nodeNum)
                         return 9041 + nodeNum;
                     }
+
+                    public int jmxPort(int nodeNum)
```
`@Override` annotation?

frankgh commented on test/distributed/org/apache/cassandra/distributed/impl/Instance.java:
```diff
@@ -754,6 +786,146 @@ else if (cluster instanceof Cluster)
+        // This prevents later attempts to unregister mbeans from failing in Cassandra code, as we're going to
+        // unregister all of them here
+        ((MBeanWrapper.DelegatingMbeanWrapper)MBeanWrapper.instance).setDelegate(new MBeanWrapper.NoOpMBeanWrapper());
+        try {
```
seems like formatting is off for this method, for example 
```suggestion
        try
        {
```

frankgh commented on test/distributed/org/apache/cassandra/distributed/impl/Instance.java:
```diff
@@ -754,6 +786,146 @@ else if (cluster instanceof Cluster)
+     * @return true, if the TCPEndpoint in the map entry was created by the InstanceClassloader instance
+     *         created by this Instance. Otherwise, false.
+     */
+    private boolean endpointCreatedByThisClassloader(Map.Entry<TCPEndpoint, LinkedList> entry) {
```
formatting?
```suggestion
    private boolean endpointCreatedByThisClassloader(Map.Entry<TCPEndpoint, LinkedList> entry)
    {
```

frankgh reviewed (commented):
Great work! I've added some minor comments, mostly around the formatting

JeetKunDoug committed:
Address review comments (formatting) and add wait for JMX to start up without blind Thread.sleep calls.
- also includes rebase on trunk and post-rebase fixes
- Clean up unnecessary modifiers in MBeanWrapper (interface members
  don't need public, static, or final modifiers)

JeetKunDoug committed:
Fix additional leak in jmx by decreasing long lease/keep-alive defaults in JMX.

Additionally, added the ability to run some code in the looper tests
that's passed from the test itself so that it can better exercise functionality.

JeetKunDoug committed:
Use new JMXUtil from dtest-api for tests.

jonmeredith commented on test/distributed/org/apache/cassandra/distributed/impl/Instance.java:
```diff
@@ -144,22 +156,30 @@
 import org.apache.cassandra.utils.logging.LoggingSupportFactory;
 import org.apache.cassandra.utils.memory.BufferPools;
 import org.apache.cassandra.utils.progress.jmx.JMXBroadcastExecutor;
+import sun.rmi.transport.tcp.TCPEndpoint;
```
should be earlier

ifesdjeen commented on .build/build-resolver.xml:
```diff
@@ -57,7 +57,7 @@
             <remoterepo id="resolver-apache" url="${artifact.remoteRepository.apache}"/>
             <!-- Snapshots are not allowed, but for feature branches they may be needed, so uncomment the below to allow snapshots to work -->
-            <!-- <remoterepo id="resolver-apache-snapshot" url="https://repository.apache.org/content/repositories/snapshots" releases="false" snapshots="true" updates="always" checksums="fail" /> -->
+            <remoterepo id="resolver-apache-snapshot" url="https://repository.apache.org/content/repositories/snapshots" releases="false" snapshots="true" updates="always" checksums="fail" />
```
Just to make sure we do not forget, this should be reverted before commit.

ifesdjeen commented on src/java/org/apache/cassandra/config/CassandraRelevantProperties.java:
```diff
@@ -247,6 +247,8 @@
     IO_NETTY_EVENTLOOP_THREADS("io.netty.eventLoopThreads"),
     IO_NETTY_TRANSPORT_ESTIMATE_SIZE_ON_SUBMIT("io.netty.transport.estimateSizeOnSubmit"),
+    /** This property indicates if the code is running under the in-jvm dtest framework */
+    IS_IN_JVM_DTEST("org.apache.cassandra.is_in_jvm_dtest"),
```
similar to howe have have `test` prefix, should we have a `dtest` prefix?

ifesdjeen commented on src/java/org/apache/cassandra/utils/MBeanWrapper.java:
```diff
@@ -36,23 +42,36 @@
         String klass = MBEAN_REGISTRATION_CLASS.getString();
         if (klass == null)
-            return new PlatformMBeanWrapper();
+            if (IS_IN_JVM_DTEST.getBoolean())
```
nit: brackets for the outer if

JeetKunDoug replied:
After discussion, we've decided to just add all of the missing single-statement curly braces to this class across all branches as it's really confusing without them at this point.

JeetKunDoug committed:
Typo fix

ifesdjeen commented on test/distributed/org/apache/cassandra/distributed/impl/Instance.java:
```diff
@@ -759,6 +789,170 @@ else if (cluster instanceof Cluster)
         initialized = true;
     }
 
+    private void startJmx() throws InterruptedException
```
this is more of a nit than an actual comment, but do you think it would be possible to move JMX-related initializers to their own class? I know we already got the jmx util class in .15, but we could also have one cassandra-locally. Most other initializations live in messaging, gossip, commit log, and other subsystems. I understand this is not necessarily matching a pattern here, but maybe we we could have a class under /impl/

JeetKunDoug replied:
Done - it really helps clean things up as well, so thanks for the suggestion.

frankgh commented on src/java/org/apache/cassandra/utils/MBeanWrapper.java:
```diff
@@ -119,17 +138,23 @@ static ObjectName create(String mbeanName, OnException onException)
         public void unregisterMBean(ObjectName mbeanName, OnException onException) {}
         public void unregisterMBean(String mbeanName, OnException onException) {}
+        public Set<ObjectName> queryNames(ObjectName name, QueryExp query) {return Collections.emptySet(); }
+        public MBeanServer getMbs() {return null; }
```
```suggestion
        public MBeanServer getMbs() { return null; }
```

JeetKunDoug committed:
Fix formatting in MBeanWrapper, move JMX to its own class

JeetKunDoug committed:
Rename prop after futher discussion

JeetKunDoug committed:
Fix shutdown bug in refactored JMX

ifesdjeen commented on src/java/org/apache/cassandra/utils/MBeanWrapper.java:
```diff
@@ -36,23 +42,46 @@
+        return getmBeanWrapper();
+    }
+
+    static MBeanWrapper getmBeanWrapper()
```
one more nit: `mBean` vs `MBean` in all other cases in this class.

ifesdjeen commented on src/java/org/apache/cassandra/utils/MBeanWrapper.java:
```diff
@@ -119,17 +154,23 @@ static ObjectName create(String mbeanName, OnException onException)
-    static class NoOpMBeanWrapper implements MBeanWrapper
+    Set<ObjectName> queryNames(ObjectName name, QueryExp query);
+
+    MBeanServer getMbs();
```
nit: maybe call it `getMBeanServer`?

JeetKunDoug committed:
Address review comments on mBean vs. MBean and rename getMbs to getMBeanServer

JeetKunDoug committed:
Register StorageServiceMBean even if gossip is disabled

JeetKunDoug committed:
Use released dtest-api artifact
